utils/result_loop.hh

```diff
@@ -251,4 +252,52 @@ result_map_reduce(Iterator begin, Iterator end, Mapper&& mapper, Initial initial
             internal::result_map_reduce_binary_adapter<left_type, right_type, Reducer>(std::move(reduce)));
 }
 
+template<typename AsyncAction, typename StopCondition>
+requires requires (StopCondition cond, AsyncAction act) {
+    { cond() } -> std::same_as<bool>;
+    { seastar::futurize_invoke(act) } -> ExceptionContainerResultFuture<>;
+}
+inline
+auto
+result_do_until(StopCondition stop_cond, AsyncAction action) {
+    // TODO: Constrain the result of act() better
+    using future_type = seastar::futurize_t<std::invoke_result_t<AsyncAction>>;
+    using result_type = typename future_type::value_type;
+    static_assert(std::is_void_v<typename result_type::value_type>,
+            "The result type of the action must be future<result<void>>");
+    for (;;) {
+        try {
+            if (stop_cond()) {
+                return seastar::make_ready_future<result_type>(bo::success());
+            }
+        } catch (...) {
+            return seastar::current_exception_as_future<result_type>();
+        }
+        auto f = seastar::futurize_invoke(action);
+        if (f.available() && !seastar::need_preempt()) {
+            if (f.failed()) {
+                return f;
+            }
+            result_type&& res = f.get();
+            if (!res) {
+                return seastar::make_ready_future<result_type>(std::move(res));
+            }
+        } else {
+            return ([] (future_type f, StopCondition stop_cond, AsyncAction action) -> seastar::future<result_type> {
+                for (;;) {
```

Comment:
Why do you have two for() loops? Is it an optimization for the case where an action is immediately ready? Does it happen often?

Reply:
Yes, it's an optimization precisely for that case. I don't know how often it happens, I just wanted to have a result-aware equivalent of the `seastar::do_until` which has the same allocation behavior, and not have to worry about allocation count increasing when I substitute `seastar::do_until` with `utils::result_do_until`.